sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import ai.rapids.cudf.{Cuda, NvtxColor, NvtxRange, Rmm}
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+
+trait MemoryLease extends AutoCloseable {
+  /**
+   * Get the amount of memory assigned to this lease
+   */
+  def leaseAmount: Long
+}
+
+/**
+ * Allows tasks to coordinate between each other to avoid
+ * oversubscribing GPU memory.
+ */
+object GpuMemoryLeaseManager extends Logging {
+  // DO NOT ACCESS DIRECTLY!  Use `getInstance` instead.
+  @volatile private var instance: GpuMemoryLeaseManager = _
+
+  private def getInstance: GpuMemoryLeaseManager = {
+    if (instance == null) {
+      GpuMemoryLeaseManager.synchronized {
+        // The instance is trying to be used before it is initialized.
+        // Since we don't have access to a configuration object here,
+        // default to only one task per GPU behavior.
```

Comment:
Do we want to complain in some way (i.e.: log or assert) if this happens?  Seems like we shouldn't be normally hitting this.  I suppose we will assert later if someone tries to initialize.  However we'd rather assert at the incorrect code point rather than some point later and wonder who incorrectly accessed the manager too early.

Reply:
I added a log message to help debug this. I was just following the pattern in the GPU Semaphore. If you want me to change it, I can, but it would probably be good to change it in both places.